eos-boot-upgrade

```diff
@@ -0,0 +1,278 @@
+
+        for part_config in disk_config['partitiontable']['partitions']:
+            part = Partition.parse(part_config, self)
+            self.partitions[part.number] = part
+
+    def get_partition_by_guid(self, guid):
+        '''Get partition by GUID / partition type'''
+        for part_num, conf in self.partitions.items():
+            if conf.type == guid:
+                return conf
+        return None
+
+    @classmethod
+    def get_booting_device(cls):
+        '''Get the booting device'''
+        root_part_device = subprocess.check_output([
+            'systemctl', 'show', '-p', 'What', 'sysroot.mount'
+            ]).decode('utf-8').strip()
+
+        root_part_device = root_part_device[5:]
+        if not root_part_device:
+            raise BootUpgradeException('Could not identify root device')
+
+        boot_disk_device = re.sub(r'p?[0-9]+$', "", root_part_device)
+        return cls(boot_disk_device)
+
+
+@contextlib.contextmanager
+def mount(device, mount_point):
+    '''Mount device on mount_point'''
+    subprocess.check_call(['mount', device, mount_point])
+    try:
+        yield mount_point
+    finally:
+        subprocess.check_call(['umount', mount_point])
+
+
+def parse_cmdline():
+    '''Parse /proc/cmdline'''
+    options = {}
+    with open('/proc/cmdline', 'r') as f:
+        for option in f.readline().strip().split():
+            fields = option.split("=", 1)
+            if len(fields) == 1:
+                options[fields[0]] = True
+            else:
+                options[fields[0]] = fields[1]
+    return options
+
+
+def get_disk():
+    '''Get the booting disk'''
+    disk = Disk.get_booting_device()
+    print('[*] Booting disk: ' + disk.device)
+
+    if disk.label != 'gpt':
+        raise BootUpgradeException('[!] Partition table is not GPT')
+
+    root_partition = disk.get_partition_by_guid('4F68BCE3-E8CD-4DB1-96E7-FBCAF984B709')
+    if not root_partition:
+        raise BootUpgradeException('[!] root partition not found (or wrong GUID)')
+
+    print('[*] Root partition: ' + root_partition.device)
+    return disk
+
+
+def get_esp(disk):
+    '''Get the ESP partition data'''
+    esp_partition = disk.get_partition_by_guid('C12A7328-F81F-11D2-BA4B-00A0C93EC93B')
+    if not esp_partition:
+        raise BootUpgradeException('[!] ESP partition not found (or wrong GUID)')
+    if not esp_partition.device.endswith('1'):
+        raise BootUpgradeException('[!] ESP is not the first partition')
+
+    print('[*] ESP partition: ' + esp_partition.device)
+    return esp_partition
+
+
+def get_dir_size(dir_path):
+    '''Get the directory size'''
+    total_size = 0
+    for (dirpath, _, filenames) in os.walk(dir_path):
+        for f in filenames:
+            fp = os.path.join(dirpath, f)
+            total_size += os.path.getsize(fp)
+    return total_size
+
+
+def check_esp_space(esp_mount, ostree_boot_efi_path):
+    '''Check that we have enough space in the ESP to copy the new data'''
```

Comment:
What's the rationale for doing this rather than just trying to copy the files over? `get_dir_size()` returns an underestimate of the size required because directories occupy space too, so we still have to handle "disk full" in the upgrade function.